espei/error_functions/chemical_potential_optimizer.py

```diff
@@ -0,0 +1,102 @@
+"""
+Author: Jonathan Siegel
+
+Contains classes and functions for optimizing at chemical potentials.
+"""
+
+import numpy as np
+from pycalphad.core.utils import point_sample
+from pycalphad.core.constants import MIN_SITE_FRACTION
+
+class PhaseConditions:
+  """
+  Contains data about a given phase at a given temperature and pressure.
+  Evaluates objectives, gradients and Hessians of the Gibbs Free energy.
+  Also contains information about the sublattices and site fractions.
+  This is passed to the optimization function as an input.
+  """
+  def __init__(self, models, phase, state_vars, obj_f, grad_f, hess_f, parameters):
+    self.state_vars = state_vars
+    self.obj_f = obj_f
+    self.grad_f = grad_f
+    self.hess_f = hess_f
+    self.phase = phase
+    self.parameters = np.array(list(parameters))
+    # Collect information about sublattices and which components they contain.
+    self.site_fraction_counts = []
+    self.site_ratios = {}
+    ind = 0
+    while True:
+      count = len([x for x in models[phase].site_fractions if x.sublattice_index == ind])
+      if count == 0:
+        break
+      self.site_fraction_counts.append(count)
+      ind += 1
+    # Generate matrix mapping site fractions to composition. Remove vacancies.
+    self.composition_conversion_matrix = \
+          np.zeros([len([x for x in models[phase].components if x.number_of_atoms > 0]), \
+                    len(models[phase].site_fractions)])
+    # Collect site ratios.
+    for i, x in enumerate(models[phase].site_fractions):
+      if x.species.number_of_atoms > 0:
+        self.site_ratios[x.sublattice_index] = models[phase].site_ratios[x.sublattice_index]
+    # Normalize the site ratios to sum to 1.
+    self.site_ratios = np.array(list(self.site_ratios.values()))
+    self.site_ratios /= np.sum(self.site_ratios)
+    for i, x in enumerate(models[phase].site_fractions):
+      if x.species.number_of_atoms > 0:
+        comp_idx = models[phase].components.index(x.species)
+        self.composition_conversion_matrix[comp_idx][i] = self.site_ratios[x.sublattice_index]
```

Comment:
Usually the site ratios are the stoichiometric coefficients for each sublattice. Is that what is meant here? The number of atoms actually means something slightly different. In CALPHAD, you can have complex species, e.g. just a pure elements Cu or O, or you could have a Cu2O complex, where the number of atoms in that complex would be 3. It's actually not related to the sublattice site ratios at all, which can most easily be found from the `sublattices` attribute of a `Phase` object:

`dbf.phases[phase_name].sublattices`

Also, before changing anything, why are these used?

Reply:
That was a hacky trick to exclude sublattices which were pure vacancies VA, which I'm not really sure what to do. I need to do this when I construct a matrix which converts site fractions into composition.

Reply:
Ah okay. The mass objective callable should convert the DOF into composition space